packages/material-ui/src/InputBase/InputBase.js

```diff
@@ -142,89 +141,131 @@ export const styles = theme => {
+  if (disabled && focused) {
+    setFocused(false);
   }
 
-  constructor(props) {
-    super(props);
-    this.isControlled = props.value != null;
-  }
+  const checkDirty = obj => {
+    if (isFilled(obj)) {
+      if (muiFormControl && muiFormControl.onFilled) {
+        muiFormControl.onFilled();
+      }
+      if (onFilled) {
+        onFilled();
+      }
+      return;
+    }
 
-  state = {
-    focused: false,
+    if (muiFormControl && muiFormControl.onEmpty) {
+      muiFormControl.onEmpty();
+    }
+    if (onEmpty) {
+      onEmpty();
+    }
   };
 
-  componentDidMount() {
-    this.checkDirty(this.isControlled ? this.props : this.inputRef);
-  }
-
-  componentDidUpdate(prevProps) {
+  React.useEffect(() => {
     // Book keep the focused state.
-    if (!prevProps.disabled && this.props.disabled) {
-      const { muiFormControl } = this.props;
+    if (disabled) {
       if (muiFormControl && muiFormControl.onBlur) {
         muiFormControl.onBlur();
       }
     }
-    if (this.isControlled) {
-      this.checkDirty(this.props);
-    } // else performed in the onChange
-  }
 
-  handleFocus = event => {
-    const { muiFormControl } = this.props;
+    checkDirty(isControlled ? props : inputRef.current);
+  });
+
+  const handleFocus = event => {
     // Fix a bug with IE 11 where the focus/blur events are triggered
     // while the input is disabled.
-    if (formControlState({ props: this.props, muiFormControl, states: ['disabled'] }).disabled) {
+    if (formControlState({ props, muiFormControl, states: ['disabled'] }).disabled) {
       event.stopPropagation();
       return;
     }
 
-    this.setState({ focused: true });
-    if (this.props.onFocus) {
-      this.props.onFocus(event);
+    setFocused(true);
+    if (onFocus) {
+      onFocus(event);
     }
 
     if (muiFormControl && muiFormControl.onFocus) {
       muiFormControl.onFocus(event);
     }
   };
 
-  handleBlur = event => {
-    this.setState({ focused: false });
-    if (this.props.onBlur) {
-      this.props.onBlur(event);
+  const handleBlur = event => {
+    setFocused(false);
+    if (onBlur) {
+      onBlur(event);
     }
 
-    const { muiFormControl } = this.props;
     if (muiFormControl && muiFormControl.onBlur) {
       muiFormControl.onBlur(event);
     }
   };
 
-  handleChange = (...args) => {
-    if (!this.isControlled) {
-      this.checkDirty(this.inputRef);
+  const handleChange = (...args) => {
+    if (!isControlled) {
+      checkDirty(inputRef.current);
     }
 
     // Perform in the willUpdate
-    if (this.props.onChange) {
-      this.props.onChange(...args);
+    if (onChange) {
+      onChange(...args);
     }
   };
 
-  handleRefInput = ref => {
-    this.inputRef = ref;
+  const handleRefInput = instance => {
+    const { inputProps } = props;
+    inputRef.current = instance;
 
     warning(
-      !ref || ref instanceof HTMLInputElement || ref.focus,
+      !instance || instance instanceof HTMLInputElement || instance.focus,
```

Comment:
https://github.com/mui-org/material-ui/pull/13934#pullrequestreview-185965047 @eps1lon @oliviertassinari How much does this warning help? I guess it would catch `<InputBase inputComponent={props => { props.inputRef("invalid"); return <div />}} />`.